docs/src/references/mappers/c8y-mapper.md

```diff
@@ -1325,3 +1325,230 @@ te/device/main///cmd/log_upload/<cmd_id>
+    MQTT Broker ->> Tedge Mapper: Telemetry message to `te/device/nested1///m/temperature`
+    Tedge Mapper ->> Tedge Mapper: Auto-register the `nested1` device as immediate child device
+    MQTT Broker ->> Tedge Mapper: Deliver child registration message on `te/device/nested1//` `{ "@parent": "device/child1//" }`
+    Tedge Mapper ->> Tedge Mapper: Ignore the child registration message as the entity is already registered
+```
+
+Here a device that was meant to be registered as a nested child device got registered as immediate child device,
+which can't be undone/corrected when the actual registration message eventually arrives.
+
+#### Initial solution thoughts
+
+One obvious solution to avoid this problem is to force the users to disable auto-registration,
+when they intend to use explicit registration even for a single entity.
+But the downside with that approach is that we force them to use explicit registration for everything,
+even the entities for which auto-registration would have been perfectly fine,
+just because a few entities required some customizations via explicit registration.
+
+The other approach would be to resort to auto-registration only when
+an explicit registration message is not delivered within a predefined window(timeout).
+But, the same timing uncertainty issues described above, applies here as well:
+what if the explicit registration message is received after the auto-registration has already happened after the timeout?
+Crashing itself isn't gonna help here either, as the entity is already auto-registered.
+
+### Final Solution
+
+The ordering problems when the mapper is live and when it is restarted would be handled differently.
+
+#### Phase 1
+
+**Live mapper**
+
+When the mapper is live, if telemetry messages for a new entity is received before its registration message,
+they are simply dropped until that registration message is processed, with the assumption that,
+telemetry data updates will repeat eventually and hence will make up for the lost ones.
+But, all non-telemetry data like twin updates, capability messages etc are cached in an in-memory "unregistered entity store",
+and will be registered in the entity store once the registration message eventually arrives.
+
+Any nested entity registration messages that arrives before their ancestors are registered
+will also be kept in the same "unregistered entity store" until all their ancestors are registered.
+
+To minimize dropping a lot of telemetry messages, they can also be cached briefly in a bounded queue,
+as in most practical cases, even if the registration messages are late, they will be delivered
+within a fair amount of time and we can avoid dropping messages in such cases.
+Keeping that queue bounded is critical to prevent it from growing indefinitely.
+
+There will also be max time limit for an unregistered entity to stay cached, after which it will be dropped.
```

Comment:
Rather than a time limit, I would use a fix-size cache evicting the oldest message when full. This is simpler to implement as no timer is required.

I would do the same for telemetry data. These caches can be small (e.g. 10 pending registrations, 100 pending telemetry data) as the point is to cope with messages that have been sent in order (root registrations < child registrations < telemetry) but received in a different order.

Reply:
The limitation with size based eviction is that the entries will remain there, consuming resources if that buffer never gets full. For example, if there are 99 large event messages sitting in the queue, with the 100th one never arriving, they'll sit there forever. That's why I was considering a time-based eviction instead, using our timer actor.